### PR TITLE
Use activerecord-import for bulk importing tracepoint records

### PR DESCRIPTION
**Pull request by gravitystorm**
Non-rigourous testing shows a significant speedup, even on ssds.

### CODE REVIEW

**Comment by gravitystorm** [comment]
One consideration here is around memory usage. The gpx.points yields tracepoints one at a time, which keeps memory usage low. But to insert them in bulk, we obviously need many tracepoints available at the same time, so we collect them up before bulk inserting.

Thoughts welcome! Do you think this will work as-is, or do we need to be more defensive about building the tracepoint array?

File: app/models/trace.rb [hunk 1]
```diff
@@ -303,9 +305,16 @@ def import
...
-      tp.save!
+      tracepoints << tp
     end
 
+    # Run the before_save and before_create callbacks, and then import them in bulk with activerecord-import
+    tracepoints.each do |tp|
+      tp.run_callbacks(:save) { false }
+      tp.run_callbacks(:create) { false }
+    end
+    Tracepoint.import(tracepoints)
```
Comment (tomhughes) [inline app/models/trace.rb]: Shouldn't this be `import!` to raise an exception on error? The old save calls were written that way?

**Comment by tomhughes** [comment]
As far as memory goes, the largest trace we have seems to have about fifteen million points, which raises a few issues - both memory as you say but also the size of the SQL if they are all inserted in one go!

Maybe we should do them in batches of a thousand or something?

File: app/models/trace.rb (hunk 1 above)
Comment (mmd-osm) [reply app/models/trace.rb]: Re. batch size: https://github.com/zdennis/activerecord-import#batching mentions some parameter to control the max number of rows per INSERT.

**Comment by mmd-osm** [comment]
I wonder if we should introduce maximum gpx file size limit to control memory consumption and processing time. 10-20 MB after decompression (and still before parsing) could be large enough for most users, and still reasonably small.

**Comment by gravitystorm** [comment]
> We also need to improve zip file handling in general here, so people can't kill the server by uploading funny zip bombs.

@mmd-osm Could you open a separate issue for this please?

File: app/models/trace.rb (hunk 1 above)
Comment (gravitystorm) [reply app/models/trace.rb]: I fixed this in 4d45a35f15c7f92c125ff070bf4757adc064ef9f

Comment (tomhughes) [reply app/models/trace.rb]: Well that stops us generating one mammoth SQL statement yes, but it still creates in-memory objects for all the points. Not sure whether that is actually an issue or not?

Comment (gravitystorm) [reply app/models/trace.rb]: Good point, but I don't know, tbh.

I can look at doing something clever to yield the points in batches too.

Comment (gravitystorm) [reply app/models/trace.rb]: OK, I've found a way to use `enum_for` to allow a simple way to pull points out of gpx.points using `.each_slice`. That means we will do multiple bulk_imports but without having more than 1,000 points in memory at one time.

**Pull request merged**